Add security extensions

gbiggs opened the pull request:
This PR is a response to [#180](https://github.com/ros2/launch_ros/pull/180). In that PR, a request was made that extensibility be added to the launch system and the `--secure` option be added as an extension. This PR provides the security extensions. The extensibility support is provided in [this PR](https://github.com/ros2/launch_ros/pull/216).

This PR adds:
- An extension to `ros2 launch` that adds the `--secure` and `--no-create-keystore` options.
- An extension to the node action from `launch_ros` that adds the necessary flags to a node's command line to start it with security.

gbiggs committed:
Add security extensions

Signed-off-by: Geoffrey Biggs <gbiggs@killbots.net>

gbiggs committed:
Ignore pycache directories

Signed-off-by: Geoffrey Biggs <gbiggs@killbots.net>

gbiggs committed:
Fix copyrights and formatting and setup for tests

Signed-off-by: Geoffrey Biggs <gbiggs@killbots.net>

gbiggs committed:
Add node action tests

Signed-off-by: Geoffrey Biggs <gbiggs@killbots.net>

gbiggs commented:
Ready for review

gbiggs committed:
Add author and maintainer information

Signed-off-by: Geoffrey Biggs <gbiggs@killbots.net>

gbiggs commented:
@clalancette Kyle from Canonical is going to take on this package and figure out who to ultimately make maintainer.

hidmic commented on ros2launch_security/option/security.py:
```diff
@@ -0,0 +1,147 @@
+                          'no keystore was provided'))
+
+
+class NonexistantKeystoreError(Exception):
```
@gbiggs typo:
```suggestion
class NonexistentKeystoreError(Exception):
```
? Or perhaps I'm unaware of this different spelling.

gbiggs replied:
Nope, just a typo! Good catch.

hidmic commented on test/test_node_action.py:
```diff
@@ -0,0 +1,136 @@
+
+class TestNodeActionSecurityExtension(unittest.TestCase):
+
+    def _assert_launch_errors(self, actions):
```
@gbiggs unused?

gbiggs replied:
Removed in ce8695a.

hidmic commented on test/test_node_action.py:
```diff
@@ -0,0 +1,136 @@
+                    'ROS_SECURITY_ENABLE': 'true',
+                },
+            ):
+                node_fixture = self._create_node()
```
@gbiggs nit^2: why not inlining `_create node` here? Every other test does so.

gbiggs replied:
At a guess (I didn't write this code), this is an incomplete refactor.

gbiggs replied:
The three occurrences are all slightly different, so I'm inclined to leave these as they are.

hidmic commented on test/test_option.py:
```diff
@@ -0,0 +1,80 @@
+    assert sros2.api._keystore.is_valid_keystore(keystore_path)
+
+    # Test that it doesn't delete a provided keystore when done
+    del res
```
@gbiggs invoking `del` on a variable simply decrements its ref count. You may want to force garbage collection for this test to test what you want it to test :) Same elsewhere.

gbiggs replied:
Fixed in db0237e.

hidmic reviewed (approved):
Overall LGTM pending green CI !

gbiggs committed:
Match changes in the extension API

Signed-off-by: Geoffrey Biggs <gbiggs@killbots.net>

gbiggs committed:
Fix spelling error

Co-authored-by: Michel Hidalgo <michel@ekumenlabs.com>

gbiggs committed:
Remove unused function

Signed-off-by: Geoffrey Biggs <gbiggs@killbots.net>

gbiggs committed:
Merge remote-tracking branch 'origin/security_extension' into security_extension

gbiggs committed:
Force garbage collection to explicitly remove objects

Signed-off-by: Geoffrey Biggs <gbiggs@killbots.net>

hidmic commented on setup.py:
```diff
@@ -0,0 +1,45 @@
+    author='Joe Example',
+    author_email='joe@example.com',
+    maintainer='Joe Example',
+    maintainer_email='joe@example.com',
```
@gbiggs not sure if Joe should be a maintainer.

gbiggs replied:
Good catch! @kyrofa was going to be maintainer of this package, but since he's left Canonical I expect he's no longer available. I don't think there is anyone else lined up yet. I will need to ask the Security WG who can be maintainer before this can be merged.

hidmic reviewed (approved):
LGTM but for one comment.

gbiggs committed:
Match changes in the extension API

Signed-off-by: Geoffrey Biggs <gbiggs@killbots.net>

hidmic commented on package.xml:
```diff
@@ -0,0 +1,30 @@
+  <maintainer email="kyle@canonical.com">Kyle Fazzari</maintainer>
+  <license>Apache License 2.0</license>
+  <author email="geoff@openrobotics.org">Geoffrey Biggs</author>
+  <author email="kyle@canonical.com">Kyle Fazzari</author>
```
@gbiggs we should update this too, if @kyrofa won't be around.

hidmic commented:
@clalancette how shall we handle running CI for this PR? If I'm not mistaken, this repository won't join the core. A supplemental repos file perhaps?

clalancette commented:
> @clalancette how shall we handle running CI for this PR? If I'm not mistaken, this repository won't join the core. A supplemental repos file perhaps?

A supplementary repos file would work.  Personally, though, I would just make a primary repos file that contains both this and  the branch for https://github.com/ros2/launch_ros/pull/216 .

clalancette commented:
Here's CI for this now that ros2/launch_ros#216 is merged:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=14237)](http://ci.ros2.org/job/ci_linux/14237/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=9035)](http://ci.ros2.org/job/ci_linux-aarch64/9035/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=11928)](http://ci.ros2.org/job/ci_osx/11928/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=14341)](http://ci.ros2.org/job/ci_windows/14341/)

clalancette commented:
This one has test failures in finding the `sros2.api._keystore` Python module.  It also needs a bit of touch-up according to @hidmic's last review.

hidmic commented:
I'll take care. Freeze is today and I want to make sure no more changes are needed in the core.

hidmic committed:
Fix failing tests

- Target sros2.keystore._keystore, not sros2.api._keystore
- Target nodl module, not launch_ros.actions.node.nodl
- Patch nodl.get_node_by_executable() properly
- Lookup enclave in Node.cmd attribute
- Add demo_nodes_py as test dependency

Signed-off-by: Michel Hidalgo <michel@ekumenlabs.com>

hidmic commented:
Alright, all issues were in this repository. I've PR'd the changes to this PR in #2. We can run CI once that gets merged into this.

gbiggs committed:
Merge pull request #2 from hidmic/security_extension

Fix failing tests

hidmic commented:
Running CI again:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=14323)](http://ci.ros2.org/job/ci_linux/14323/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=9111)](http://ci.ros2.org/job/ci_linux-aarch64/9111/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12008)](http://ci.ros2.org/job/ci_osx/12008/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=14427)](http://ci.ros2.org/job/ci_windows/14427/)

hidmic commented:
Argh, was missing the `nodl` repo:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=14326)](http://ci.ros2.org/job/ci_linux/14326/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=9113)](http://ci.ros2.org/job/ci_linux-aarch64/9113/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12012)](http://ci.ros2.org/job/ci_osx/12012/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=14431)](http://ci.ros2.org/job/ci_windows/14431/)

hidmic commented:
That should get separately released into Rolling though.

hidmic commented:
@gbiggs this is good to go. Have you managed to figure out who's going to be the maintainer?

gbiggs commented:
Not yet. The last security WG meeting was at 2AM. I think I'll send a message instead.

kyrofa commented:
The new maintainer should be @Arnatious.

gbiggs commented:
@kyrofa Can you supply an email address for the maintainer line?

kyrofa commented on package.xml:
```diff
@@ -0,0 +1,31 @@
+  <name>ros2launch_security</name>
+  <version>0.1.0</version>
+  <description>Security extensions for ros2 launch</description>
+  <maintainer email="kyle@canonical.com">Kyle Fazzari</maintainer>
```
```suggestion
  <maintainer email="ted.kern@canonical.com">Ted Kern</maintainer>
```

kyrofa commented on package.xml:
```diff
@@ -0,0 +1,31 @@
+  <maintainer email="kyle@canonical.com">Kyle Fazzari</maintainer>
+  <license>Apache License 2.0</license>
+  <author email="geoff@openrobotics.org">Geoffrey Biggs</author>
+  <author email="kyle@canonical.com">Kyle Fazzari</author>
```
```suggestion
  <author email="ted.kern@canonical.com">Ted Kern</author>
```

gbiggs committed:
Set maintainer info

Co-authored-by: Kyle Fazzari <kyrofa+github@gmail.com>

gbiggs committed:
Add author info

Co-authored-by: Kyle Fazzari <kyrofa+github@gmail.com>

gbiggs commented:
@hidmic Thanks to @kyrofa we have a maintainer now. The `package.xml` has been updated with their information.

Arnatious commented:
Just confirming I'm taking over maintaining

hidmic commented:
@gbiggs @Arnatious what's pending to move forward with this?

gbiggs commented:
Nothing as far as I know, so I'm going to go ahead and merge it.
